:arrow_up: Upgrade Zarr to v3+ Compatibility

measty opened the pull request:
## ✅ Summary

- Enables compatibility with Zarr v3+
- Simplifies error handling across Zarr versions
- Removes deprecated storage backends and APIs
- Lays groundwork for full Zarr v3 migration

This PR updates TIAToolbox to support **Zarr v3+**, addressing breaking changes introduced in Zarr ≥ 3.0 and aligning internal APIs, storage handling, and error handling accordingly.

---

Zarr v3 removed `zarr.errors.FSPathExistNotDir`. In Zarr ≥ 3.0, equivalent failure modes now raise standard Python exceptions (e.g. `FileNotFoundError`).

To avoid fragile version-specific logic:
- Removed checks for specific Zarr exception types
- Replaced them with generic exception handling that works consistently across Zarr versions

---

## 🔧 Key changes

### Dependency updates
- 🆙 Upgrade to:
  - `zarr >= 3.0.8`
  - `tifffile >= 2025.5.21`

### Zarr API & storage migration
- Refactor `zarr.core.Array` → `zarr.Array`
- Replace `zarr.DirectoryStore` with `zarr.storage.LocalStore`
- Remove `ngff.sqlitestore` (Zarr v3 no longer supports `zarr.SQLiteStore`)
- Replace `canvas_zarr.store.path` with `canvas_zarr.store.root`
- Ensure Zarr paths explicitly use string keys (`key / str(i)` instead of `key / i`)

### Dask & array handling
- Remove unsupported `object_codec` usage in `dask.to_zarr`
- Update `zarr.resize` calls to use tuple inputs
- Replace `output.items()` with `output.members()`

---

## ⚠️ Known limitations & follow‑ups

Zarr v3 removes or significantly changes several APIs still used by TIAToolbox, including:
- `zarr.LRUStoreCache`
- `zarr.DirectoryStore`
- `zarr.SQLiteStore`

While this PR resolves immediate incompatibilities and enables Zarr v3 usage, **additional refactoring will be required** to fully replace or redesign components that depend on these removed features.

In addition, **full compatibility with `tifffile` depends on upstream Zarr v3 support**, and some workflows may remain limited until that stabilises.

📖 For a detailed overview of breaking changes, see the official Zarr v3 migration guide:  
https://zarr.readthedocs.io/en/latest/user-guide/v3_migration.html

---

measty committed:
fix zarr checking

measty committed:
use cross-version syntax for zarr.group

measty committed:
more zarr v3 changes

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:pushpin: Remove Pin from Zarr

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed commented:
Zarr v3 is not compatible with tifffile https://github.com/cgohlke/tifffile/issues/282

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed commented:
Zarr 3 is supported by tifffile. However, zarr 3 only supports Python 3.11+.

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:pushpin: Pin `zarr` and `tifffile`

shaneahmed committed:
:pushpin: Pin `tifffile`

shaneahmed committed:
:pushpin: Update Python Versions
- Drop Python 3.10 support
- Add Python 3.14 support

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shaneahmed committed:
:bug: Fix setup.py to include 3.14

shaneahmed committed:
Merge remote-tracking branch 'origin/dev-remove-python3.10' into dev-remove-python3.10

shaneahmed committed:
:fire: Remove openslide from requirements.conda.yml

shaneahmed committed:
:fire: Remove Python 3.14
- Support for Python 3.14 needs a separate PR

shaneahmed committed:
Merge branch 'dev-remove-python3.10' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:bug: Fix Semantic Segmentor
- Remove `object_codec` in dask.to_zarr
- Refactor `zarr.core.Array` to `zarr.Array`
- Refactor `canvas_zarr.store.path` to `canvas_zarr.store.root`
- Refactor `zarr.DirectoryStore` to `zarr.storage.LocalStore`
- tuple input for zarr.resize
- Replace `output.items()` with `output.members()`

shaneahmed committed:
:bug: Fix misc.py

shaneahmed committed:
:bug: Fix `wsireader.py`
- Use `CacheStore` for zarr v3

shaneahmed committed:
:white_check_mark: Force skip test

shaneahmed committed:
:bug: Fix NGFF Reader

shaneahmed committed:
:bug: Fix NGFF Reader
- Remove sqlite tests

shaneahmed committed:
:white_check_mark: Use FsspecStore for remote files.

shaneahmed committed:
Merge remote-tracking branch 'origin/fix-zarr-check' into fix-zarr-check

shaneahmed commented:
@aacic Current implementation of `FsspecJsonWSIReader` is not compatible with Zarr V3. I have tested multiple options but they fail. Do you have any suggestions?

https://github.com/TissueImageAnalytics/tiatoolbox/blob/750ca026df82839e1e286b2e25d17ef53aa55a85/tiatoolbox/wsicore/wsireader.py#L4392-L4459

shaneahmed committed:
:recycle: Update how `register_codec` is used

shaneahmed committed:
:white_check_mark: Add support to read s3 using NGFFWSIReader

shaneahmed committed:
:pencil2: Fix typos

aacic commented:
> @aacic Current implementation of `FsspecJsonWSIReader` is not compatible with Zarr V3. I have tested multiple options, but they fail. Do you have any suggestions?
> 
> https://github.com/TissueImageAnalytics/tiatoolbox/blob/750ca026df82839e1e286b2e25d17ef53aa55a85/tiatoolbox/wsicore/wsireader.py#L4392-L4459

Please give me some time to look into this. I'll try to look at this next week. I don't know it from the top of my head.

shaneahmed committed:
:bug: Fix `nucleus_detector.py`

shaneahmed committed:
:bug: Zarr uses str indexing instead of int

shaneahmed committed:
:white_check_mark: Test with np full

shaneahmed committed:
:white_check_mark: Fix mtsegmentor patches and tiles_no_metadata

shaneahmed committed:
:white_check_mark: Fix mtsegmentor patches

shaneahmed committed:
:white_check_mark: Fix test_single_task_mtsegmentor

shaneahmed committed:
:white_check_mark: Fix test_wsi_mtsegmentor_correct_nonsquare_shape and test_wsi_mtsegmentor_zarr

shaneahmed committed:
:white_check_mark: Fix test_wsi_segmentor_annotationstore

shaneahmed commented:
All the errors have been fixed except `FsspecJsonWSIReader`.

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:white_check_mark: Fix test_micronet_output

shaneahmed commented:
> All the errors have been fixed except `FsspecJsonWSIReader`.

The updates to engines output breaks, `test_hovernet_on_box` which also needs to be fixed.

aacic commented:
> All the errors have been fixed except `FsspecJsonWSIReader`.

@shaneahmed  I was able to make it work locally. I need some more time to verify the solution, and I'll try to submit a PR sometime next week. I'll keep you posted.

shaneahmed commented:
> > All the errors have been fixed except `FsspecJsonWSIReader`.
> 
> @shaneahmed I was able to make it work locally. I need some more time to verify the solution, and I'll try to submit a PR sometime next week. I'll keep you posted.

Thank you @aacic That would be wonderful.

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

# Conflicts:
#	requirements/requirements.txt

shaneahmed commented:
> > All the errors have been fixed except `FsspecJsonWSIReader`.
> 
> The updates to engines output breaks, `test_hovernet_on_box` which also needs to be fixed.

@measty Please would you be able to check the error with ` test_hovernet_on_box`? This is probably due to updates in the way new zarr output is saved.

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

aacic committed:
:arrow_up: `FsspecJsonWSIReader` Zarr 3 Fix  (#1049)

* Zarr 3 fix.

* Fix multilayer image `FsspecJsonWSIReader` support.

JiwaniZakir reviewed (commented):
The padding-removal logic added in `assert_annotation_store_patch_output` and `assert_qupath_json_patch_output` uses `np.iinfo(contours.dtype).min` as the sentinel value, but `np.iinfo` only accepts integer dtypes — if `contours.dtype` is ever a float (e.g., `float32` or `float64`), this will raise a `ValueError` at runtime. A safer approach would be to guard with `np.iinfo` only when `np.issubdtype(contours.dtype, np.integer)` and fall back to `np.finfo` or a different sentinel otherwise.

The pervasive switch from integer to string keys (e.g., `output_zarr["x"][0]` → `output_zarr["x"]["0"]` in `test_nucleus_detection_engine.py` and the `str(i)` conversion in `assert_output_lengths`) reflects a real zarr v3 API change, but it's worth verifying this pattern is consistently applied in the production code paths as well — the tests only cover the assertion helpers, and any production code that still uses integer indexing into zarr groups would silently fail or return wrong results.

The switch from `output_.items()` to `output_.members()` in `test_semantic_segmentor.py` is correct for zarr v3, but the truncated diff makes it unclear whether similar `items()` calls in production code (not just tests) have been updated — a quick grep for `.items()` on zarr group objects in the main source would be worth confirming.

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

Jiaqi-Lv committed:
fix mypy errors

shaneahmed committed:
:bug: Fix missing mask for contours

shaneahmed committed:
:hammer: Use `skip` for follow imports

Jiaqi-Lv committed:
fix mypy type errors

shaneahmed committed:
:hammer: Mark s3 test for NGFF as expected to fail
- This test is expected to fail depending on external source being available.

shaneahmed committed:
:bug: Fix deepsource error

shaneahmed committed:
:bug: Fix deepsource error cyclomatic complexity too high.

shaneahmed commented on tests/engines/test_multi_task_segmentor.py:
```diff
@@ -1205,7 +1209,10 @@ def assert_output_lengths(
     for field in fields:
         for i, expected in enumerate(expected_counts):
-            assert len(output[field][i]) == expected, f"{field}[{i}] mismatch"
+            idx = str(i) if isinstance(output[field], (zarr.Array, zarr.Group)) else i
```
```suggestion
            idx = str(i) if isinstance(output[field], zarr.Group) else i
```

shaneahmed commented on tests/engines/test_multi_task_segmentor.py:
```diff
@@ -1263,11 +1270,21 @@ def assert_output_equal(
-            right = output_b[field][i_b]
+            i_a_ = (
+                str(i_a)
+                if isinstance(output_a[field], (zarr.Array, zarr.Group))
```
```suggestion
                if isinstance(output_a[field], zarr.Group)
```

shaneahmed commented on tests/engines/test_multi_task_segmentor.py:
```diff
@@ -1263,11 +1270,21 @@ def assert_output_equal(
+            )
+            i_b_ = (
+                str(i_b)
+                if isinstance(output_b[field], (zarr.Array, zarr.Group))
```
```suggestion
                if isinstance(output_b[field], zarr.Group)
```

shaneahmed committed:
:bug: Fix instance test with zarr.Array

shaneahmed committed:
:white_check_mark: Add tests for coverage

shaneahmed committed:
:white_check_mark: Add tests for wsireader coverage

shaneahmed committed:
:white_check_mark: Add tests for multi_task_segmentor coverage

shaneahmed committed:
:fire: Remove dtype object check

shaneahmed committed:
:bug: Fix "store" attribute error with dictionary

shaneahmed committed:
:bulb: Address Co-Pilot comments

shaneahmed committed:
:bulb: Address Co-Pilot comments

shaneahmed committed:
:white_check_mark: Add tests to improve coverage

shaneahmed reviewed (approved):
I have tested the engines locally and they work well. I have also addressed Co-Pilots comments.

shaneahmed committed:
:white_check_mark: Add tests to improve coverage

shaneahmed committed:
:technologist: Address quality check issues

shaneahmed commented on tests/engines/test_multi_task_segmentor.py:
```diff
@@ -1534,3 +1564,248 @@ def test_cli_model_single_file(remote_sample: Callable, track_tmp_path: Path) ->
+    ) -> dict[str, Any]:
+        return qupath_json
+
+    import tiatoolbox.models.engine.multi_task_segmentor as mts  # noqa: PLC0415
```
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/904/commits/ccb0a14dcfd144d26cc210c0c54f42bfcab6f6e5

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:bug: Fix pip install workflow

shaneahmed committed:
Merge branch 'develop' into fix-zarr-check

shaneahmed committed:
:bug: Replace `create_dataset` with `create_array`

shaneahmed committed:
:bug: Fix test_clear_zarr

shaneahmed committed:
:bug: Fix chunksize 0 unsupported by zarr v3.2.0+
